src/srutils/dataclasses.py

```diff
@@ -299,45 +302,43 @@ def prepare_wrapped_value(
                 }.get(str(value).lower(), value)
         return bool(value)
 
-    elif type_ == "str":
+    elif type_name == "str":
         if isinstance(value, Collection) and not isinstance(value, str):
-            raise error(value, type_, "is a collection")
+            raise error(value, type_name, "is a collection")
         return str(value)
 
-    elif type_ in ["int", "float"]:
+    elif type_name in ["int", "float"]:
         try:
-            return {"int": int, "float": float}[type_](value)
+            return {"int": int, "float": float}[type_name](value)
         except (TypeError, ValueError) as e:
-            raise error(value, type_) from e
+            raise error(value, type_name) from e
 
-    elif type_.startswith("typing.Union["):
-        inner_types = split_outside_parens(
-            type_[len("typing.Union[") : -len("]")], ", ", parens="[]"
-        )
+    elif "Union" in type_name:
+        inner_types = get_args(type_)
         if "NoneType" in inner_types:
             inner_types.remove("NoneType")
             inner_types.insert(0, "NoneType")
         for inner_type in inner_types:
-            if has_same_type(value, inner_type):
+            if type(value) == inner_type:
                 return value
         for inner_type in inner_types:
             try:
                 return cast_value(inner_type, value, **kwargs)
             except IncompatibleTypesError:
                 pass
-        raise error(value, type_, f"no compatible inner type: {inner_types}")
+        raise error(value, type_name, f"no compatible inner type: {inner_types}")
 
-    elif type_.startswith("typing.Optional["):
+    elif "Optional" in type_name:
         if value in [None, "None"]:
             return None
-        inner_type = type_[len("typing.Optional[") : -len("]")]
+        inner_type = [i for i in get_args(type_) if not i == type(None)][0]
```

Comment:
If a single type is expected, the tuple unpacking mechanism can be used to verify this:
```suggestion
        [inner_type] = (i for i in get_args(type_) if not i == type(None))
```